crates/sui-core/src/authority_server.rs

```diff
@@ -452,6 +473,120 @@ impl ValidatorService {
+
+        // Check system overload
+        let overload_check_res = self.state.check_system_overload(
+            &consensus_adapter,
+            transaction.data(),
+            state.check_system_overload_at_signing(),
+        );
+        if let Err(error) = overload_check_res {
+            metrics
+                .num_rejected_tx_during_overload
+                .with_label_values(&[error.as_ref()])
+                .inc();
+            return Err(error.into());
+        }
+
+        let _handle_tx_metrics_guard = metrics.handle_transaction_v2_latency.start_timer();
+
+        let tx_verif_metrics_guard = metrics.tx_verification_latency.start_timer();
+        let transaction = epoch_store.verify_transaction(transaction).tap_err(|_| {
+            metrics.signature_errors.inc();
+        })?;
+        drop(tx_verif_metrics_guard);
+
+        // Enable Trace Propagation across spans/processes using tx_digest
+        let tx_digest = transaction.digest();
+        let span = error_span!("validator_state_process_tx_v2", ?tx_digest);
+
+        let tx_status = state
+            .handle_transaction_v2(&epoch_store, transaction.clone())
+            .instrument(span)
+            .await
+            .tap_err(|e| {
+                if let SuiError::ValidatorHaltedAtEpochEnd = e {
+                    metrics.num_rejected_tx_in_epoch_boundary.inc();
+                }
+            })?;
+        if let Some((
+            _sender_signed_data,
+            // TODO(fastpath): Suppress duplicate transaction submission in consensus
+            // adapter, if not already done. (If we get back `TransactionStatus::Signed``
+            // here we still need to proceed with submission logic, because the previous
+            // RPC might have been dropped after signing but before submission.)
+            TransactionStatus::Executed(_sign_info, signed_effects, events),
+        )) = tx_status
+        {
+            let input_objects = include_input_objects
+                .then(|| state.get_transaction_input_objects(signed_effects.data()))
+                .and_then(Result::ok);
+            let output_objects = include_output_objects
+                .then(|| state.get_transaction_output_objects(signed_effects.data()))
+                .and_then(Result::ok);
+
+            return Ok((
+                tonic::Response::new(HandleTransactionResponseV2 {
+                    effects: signed_effects.into_data(),
+                    events: include_events.then_some(events),
+                    input_objects,
+                    output_objects,
+                    auxiliary_data: None, // We don't have any aux data generated presently
+                }),
+                Weight::zero(),
+            ));
+        }
+
+        let _latency_metric_guard = metrics.handle_transaction_consensus_latency.start_timer();
+        let span = error_span!("handle_transaction_v2", tx_digest = ?transaction.digest());
+        self.handle_submit_to_consensus(
```

Comment:
Btw let's guard this or the entire handler with a protocol config flag, e.g. `mysticeti_fastpath`? Otherwise we may accidentally expose production validators to dos or crash before the logic is complete.

Reply:
Good idea